lib/Sema/SemaBounds.cpp

```diff
@@ -345,3 +345,86 @@ void Sema::CheckTopLevelBoundsDecls(VarDecl *D) {
   if (!D->isLocalVarDeclOrParm())
     CheckBoundsDeclarations(*this).TraverseVarDecl(D);
 }
+
+
+namespace {
+  class NonModifiyingExprSema : public RecursiveASTVisitor<NonModifiyingExprSema> {
+
+  private:
+    // Represents which kind of modifying expression we have found
+    enum ModifyingExprKind {
+      MEK_Assign,
+      MEK_Increment,
+      MEK_Decrement,
+      MEK_Call,
+      MEK_Comma,
+      MEK_Volatile
+    };
+
+  public:
+    NonModifiyingExprSema(Sema &S) : S(S) {}
+
+    // Assignments are of course modifying
+    bool VisitBinAssign(BinaryOperator* E) {
+      addError(E, MEK_Assign);
+      return false;
```

Comment:
We should not terminate the walk early if we find an error.   A design goal for the compiler is to find as many errors as possible in one compilation.  Compilations can be expensive and build systems can be slow, so this improves programmer productivity.    The only reason to cut off traversal would be if we find ourselves in a situation that would later generate spurious errors or in which we can make no further meaningful progress.